File: packages/voltaire-zig/src/jsonrpc/eth/coinbase/eth_coinbase.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Returns the client coinbase address.
///
/// Example:
/// Result: "0xfe3b557e8fb62b89f4916b721be55ceb828dbd73"
///
/// Implements the `eth_coinbase` JSON-RPC method.
pub const EthCoinbase = @This();

/// The JSON-RPC method name
pub const method = "eth_coinbase";

/// Parameters for `eth_coinbase`
pub const Params = struct {
    pub fn jsonStringify(self: Params, jws: *std.json.Stringify) !void {
        _ = self;
        try jws.write(.{});
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        _ = allocator;
        _ = source;
        _ = options;
        return Params{};
    }
};

/// Result for `eth_coinbase`
pub const Result = struct {
    /// hex encoded address
    value: types.Address,

    pub fn jsonStringify(self: Result, jws: *std.json.Stringify) !void {
        try jws.write(self.value);
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        return Result{
            .value = try std.json.innerParseFromValue(types.Address, allocator, source, options),
        };
    }
};

File: packages/voltaire-ts/src/jsonrpc/eth/coinbase/eth_coinbase.go
package eth

import (
	"encoding/json"
	"fmt"

	"github.com/ethereum/execution-apis/types"
)

// Returns the client coinbase address.
//
// Example:
// Result: "0xfe3b557e8fb62b89f4916b721be55ceb828dbd73"
//
// Implements the eth_coinbase JSON-RPC method.

// Method is the JSON-RPC method name
const Method = "eth_coinbase"

// Params represents the parameters for eth_coinbase
type Params struct {
}

// MarshalJSON implements json.Marshaler for Params.
// JSON-RPC 2.0 uses positional array parameters.
func (p Params) MarshalJSON() ([]byte, error) {
	return json.Marshal([]interface{}{
	})
}

// UnmarshalJSON implements json.Unmarshaler for Params.
func (p *Params) UnmarshalJSON(data []byte) error {
	var arr []json.RawMessage
	if err := json.Unmarshal(data, &arr); err != nil {
		return err
	}

	if len(arr) != 0 {
		return fmt.Errorf("expected 0 parameters, got %d", len(arr))
	}

	return nil
}

// Result represents the result for eth_coinbase
//
// hex encoded address
type Result struct {
	Value types.Address `json:"-"`
}

// MarshalJSON implements json.Marshaler for Result.
func (r Result) MarshalJSON() ([]byte, error) {
	return json.Marshal(r.Value)
}

// UnmarshalJSON implements json.Unmarshaler for Result.
func (r *Result) UnmarshalJSON(data []byte) error {
	return json.Unmarshal(data, &r.Value)
}
